diff --git a/BUILD.bazel b/BUILD.bazel
index 95e0ffbca88a..b965e3e40815 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -769,6 +769,7 @@ _DB845C_MODULE_OUTS = [
     "drivers/net/wireless/ath/ath11k/ath11k_pci.ko",
     "drivers/net/wireless/ath/ath12k/ath12k.ko",
     "drivers/nvmem/nvmem_qfprom.ko",
+    "drivers/nxp/nfc/nxpnfc_i2c.ko",
     "drivers/pci/pwrctrl/pci-pwrctrl-pwrseq.ko",
     "drivers/phy/phy-snps-eusb2.ko",
     "drivers/phy/qualcomm/phy-qcom-edp.ko",
diff --git a/arch/arm64/boot/dts/qcom/sdm845-db845c.dts b/arch/arm64/boot/dts/qcom/sdm845-db845c.dts
index 5147d6d3cc26..f81ce4755eb3 100644
--- a/arch/arm64/boot/dts/qcom/sdm845-db845c.dts
+++ b/arch/arm64/boot/dts/qcom/sdm845-db845c.dts
@@ -265,8 +265,8 @@ vreg_s4a_1p8: vreg-s4a-1p8 {
 		compatible = "regulator-fixed";
 		regulator-name = "vreg_s4a_1p8";
 
-		regulator-min-microvolt = <1800000>;
-		regulator-max-microvolt = <1800000>;
+		regulator-min-microvolt = <1850000>;
+		regulator-max-microvolt = <1850000>;
 		regulator-always-on;
 	};
 
@@ -519,14 +519,22 @@ lt9611_out: endpoint {
 
 &i2c11 {
 	/* On Low speed expansion */
-	clock-frequency = <100000>;
+	clock-frequency = <1000000>;
 	status = "okay";
+	nxpnfc@28 {
+		compatible = "nxp,nxpnfc";
+		reg = <0x28>;
+		nxp,nxpnfc-irq = <&tlmm 49 00>;
+		nxp,nxpnfc-ven = <&tlmm 50 00>;
+		nxp,nxpnfc-fw-dwnld = <&tlmm 51 00>;
+		pinctrl-names = "default";
+	};
 };
 
 &i2c14 {
 	/* On Low speed expansion */
 	clock-frequency = <100000>;
-	status = "okay";
+	status = "disabled";
 };
 
 &mdss {
diff --git a/arch/arm64/boot/dts/qcom/sdm845.dtsi b/arch/arm64/boot/dts/qcom/sdm845.dtsi
index 13c9515260ef..591be77253b6 100644
--- a/arch/arm64/boot/dts/qcom/sdm845.dtsi
+++ b/arch/arm64/boot/dts/qcom/sdm845.dtsi
@@ -2939,6 +2939,7 @@ qup_i2c13_default: qup-i2c13-default-state {
 			qup_i2c14_default: qup-i2c14-default-state {
 				pins = "gpio33", "gpio34";
 				function = "qup14";
+				bias-high-impedance;
 			};
 
 			qup_i2c15_default: qup-i2c15-default-state {
diff --git a/arch/arm64/configs/db845c_gki.fragment b/arch/arm64/configs/db845c_gki.fragment
index 092da33fe903..670ac71b0e00 100644
--- a/arch/arm64/configs/db845c_gki.fragment
+++ b/arch/arm64/configs/db845c_gki.fragment
@@ -5,6 +5,7 @@ CONFIG_SCSI_UFS_QCOM=m
 CONFIG_INPUT_PM8941_PWRKEY=m
 CONFIG_SERIAL_MSM=m
 CONFIG_I2C_QCOM_GENI=m
+CONFIG_NXP_NFC_I2C=m
 CONFIG_I2C_QUP=m
 CONFIG_PINCTRL_MSM=m
 CONFIG_PINCTRL_QCOM_SPMI_PMIC=m
diff --git a/drivers/Kconfig b/drivers/Kconfig
index 4915a63866b0..1c03d6320d63 100644
--- a/drivers/Kconfig
+++ b/drivers/Kconfig
@@ -251,4 +251,6 @@ source "drivers/hte/Kconfig"
 
 source "drivers/cdx/Kconfig"
 
+source "drivers/nxp/Kconfig"
+
 endmenu
diff --git a/drivers/Makefile b/drivers/Makefile
index 8e1ffa4358d5..ea7381f5c0bf 100644
--- a/drivers/Makefile
+++ b/drivers/Makefile
@@ -197,3 +197,4 @@ obj-$(CONFIG_DPLL)		+= dpll/
 
 obj-$(CONFIG_DIBS)		+= dibs/
 obj-$(CONFIG_S390)		+= s390/
+obj-y				    += nxp/
diff --git a/drivers/i2c/busses/i2c-qcom-geni.c b/drivers/i2c/busses/i2c-qcom-geni.c
index bfb352b04902..00ac6ad7db7c 100644
--- a/drivers/i2c/busses/i2c-qcom-geni.c
+++ b/drivers/i2c/busses/i2c-qcom-geni.c
@@ -150,7 +150,7 @@ struct geni_i2c_clk_fld {
 static const struct geni_i2c_clk_fld geni_i2c_clk_map_19p2mhz[] = {
 	{ I2C_MAX_STANDARD_MODE_FREQ, 7, 10, 12, 26 },
 	{ I2C_MAX_FAST_MODE_FREQ, 2,  5, 11, 22 },
-	{ I2C_MAX_FAST_MODE_PLUS_FREQ, 1, 2,  8, 18 },
+	{ I2C_MAX_FAST_MODE_PLUS_FREQ, 1, 3,  6, 15 },
 	{}
 };
 
@@ -421,29 +421,29 @@ static int geni_i2c_rx_one_msg(struct geni_i2c_dev *gi2c, struct i2c_msg *msg,
 {
 	dma_addr_t rx_dma = 0;
 	unsigned long time_left;
-	void *dma_buf;
+	void *dma_buf = NULL;
 	struct geni_se *se = &gi2c->se;
 	size_t len = msg->len;
 	struct i2c_msg *cur;
-
-	dma_buf = gi2c->no_dma ? NULL : i2c_get_dma_safe_msg_buf(msg, 32);
+	//Enabled FIFO mode explicitly
+	/*dma_buf = gi2c->no_dma ? NULL : i2c_get_dma_safe_msg_buf(msg, 32);
 	if (dma_buf)
 		geni_se_select_mode(se, GENI_SE_DMA);
-	else
+	else*/
 		geni_se_select_mode(se, GENI_SE_FIFO);
 
 	writel_relaxed(len, se->base + SE_I2C_RX_TRANS_LEN);
 	geni_se_setup_m_cmd(se, I2C_READ, m_param);
-
-	if (dma_buf && geni_se_rx_dma_prep(se, dma_buf, len, &rx_dma)) {
+	//Enabled FIFO mode explicitly
+	/*if (dma_buf && geni_se_rx_dma_prep(se, dma_buf, len, &rx_dma)) {
 		geni_se_select_mode(se, GENI_SE_FIFO);
 		i2c_put_dma_safe_msg_buf(dma_buf, msg, false);
 		dma_buf = NULL;
-	} else {
+	} else {*/
 		gi2c->xfer_len = len;
 		gi2c->dma_addr = rx_dma;
 		gi2c->dma_buf = dma_buf;
-	}
+	// }
 
 	cur = gi2c->cur;
 	time_left = wait_for_completion_timeout(&gi2c->done, XFER_TIMEOUT);
@@ -460,15 +460,15 @@ static int geni_i2c_tx_one_msg(struct geni_i2c_dev *gi2c, struct i2c_msg *msg,
 {
 	dma_addr_t tx_dma = 0;
 	unsigned long time_left;
-	void *dma_buf;
+	void *dma_buf = NULL;
 	struct geni_se *se = &gi2c->se;
 	size_t len = msg->len;
 	struct i2c_msg *cur;
-
-	dma_buf = gi2c->no_dma ? NULL : i2c_get_dma_safe_msg_buf(msg, 32);
+	//Enabled FIFO mode explicitly
+	/*dma_buf = gi2c->no_dma ? NULL : i2c_get_dma_safe_msg_buf(msg, 32);
 	if (dma_buf)
 		geni_se_select_mode(se, GENI_SE_DMA);
-	else
+	else*/
 		geni_se_select_mode(se, GENI_SE_FIFO);
 
 	writel_relaxed(len, se->base + SE_I2C_TX_TRANS_LEN);
diff --git a/gki/aarch64/symbols/db845c b/gki/aarch64/symbols/db845c
index 9e3407d8b56b..6f6aade3b631 100644
--- a/gki/aarch64/symbols/db845c
+++ b/gki/aarch64/symbols/db845c
@@ -1776,6 +1776,23 @@
 # required by nvmem_qfprom.ko
   devm_nvmem_register
 
+#required by nxp nxpnfc_i2c.ko
+  __dynamic_pr_debug
+  cdev_add
+  cdev_del
+  class_create
+  class_destroy
+  device_create
+  device_destroy
+  gpio_free
+  gpio_request
+  gpiod_direction_input
+  gpiod_get_raw_value
+  gpio_to_desc
+  gpiod_direction_output_raw
+  gpiod_set_raw_value
+  of_get_named_gpio
+
 # required by ohci-hcd.ko
   dma_pool_create_node
   dma_pool_destroy
