diff --git a/db845c.mk b/db845c.mk
index c83da69..ce912fc 100644
--- a/db845c.mk
+++ b/db845c.mk
@@ -5,6 +5,7 @@ $(call inherit-product, device/linaro/dragonboard/shared/graphics/drm_hwcomposer
 $(call inherit-product, device/linaro/dragonboard/shared/graphics/mesa/device.mk)
 $(call inherit-product, device/linaro/dragonboard/shared/graphics/minigbm_msm/device.mk)
 $(call inherit-product, device/linaro/dragonboard/db845c/device.mk)
+$(call inherit-product, vendor/nxp/nfc/device-nfc.mk)
 
 #Boot with Android boot image header v3
 ifeq ($(TARGET_KERNEL_USE),5.4)
diff --git a/db845c/BoardConfig.mk b/db845c/BoardConfig.mk
index ed9264c..440765b 100644
--- a/db845c/BoardConfig.mk
+++ b/db845c/BoardConfig.mk
@@ -1,4 +1,5 @@
 include device/linaro/dragonboard/BoardConfigCommon.mk
+include vendor/nxp/nfc/BoardConfigNfc.mk
 
 # Primary Arch
 TARGET_ARCH := arm64
diff --git a/device-common.mk b/device-common.mk
index 960c2ae..5637efb 100644
--- a/device-common.mk
+++ b/device-common.mk
@@ -16,6 +16,7 @@
 
 TARGET_KERNEL_DIR ?= device/linaro/dragonboard-kernel/android-$(TARGET_KERNEL_USE)
 TARGET_MODS := $(wildcard $(TARGET_KERNEL_DIR)/*.ko)
+DB845C_ONLY_VENDOR := $(wildcard $(TARGET_KERNEL_DIR)/nxpnfc_spi.ko)
 
 BOARD_DO_NOT_STRIP_VENDOR_RAMDISK_MODULES := true
 BOARD_DO_NOT_STRIP_GENERIC_RAMDISK_MODULES := true
@@ -23,13 +24,16 @@ ifeq ($(TARGET_SDCARD_BOOT), true)
   # UFS module filename varies from ufs_qcom.ko to ufs-qcom.ko across different kernel versions
   BOARD_VENDOR_KERNEL_MODULES := $(wildcard $(TARGET_KERNEL_DIR)/ufs*qcom.ko)
   BOARD_GENERIC_RAMDISK_KERNEL_MODULES := $(filter-out $(BOARD_VENDOR_KERNEL_MODULES),$(TARGET_MODS))
+  BOARD_GENERIC_RAMDISK_KERNEL_MODULES += $(filter-out $(DB845C_ONLY_VENDOR),$(TARGET_MODS))
 else ifeq ($(TARGET_USES_LMP), true)
     include device/linaro/dragonboard/shared/utils/dlkm_loader/vendor.modules.list.mk
     BOARD_VENDOR_KERNEL_MODULES := $(patsubst %,$(TARGET_KERNEL_DIR)/%,$(VENDOR_DLKM_KERNEL_MODULES_LIST))
-    BOARD_VENDOR_RAMDISK_KERNEL_MODULES := $(patsubst %,$(TARGET_KERNEL_DIR)/%,$(VENDOR_RAMDISK_KERNEL_MODULES_LIST))
+    BOARD_VENDOR_KERNEL_MODULES += $(DB845C_ONLY_VENDOR)
+    BOARD_VENDOR_RAMDISK_KERNEL_MODULES := $(filter-out $(DB845C_ONLY_VENDOR),$(TARGET_MODS))
     BOARD_SYSTEM_KERNEL_MODULES := $(filter-out $(BOARD_VENDOR_KERNEL_MODULES) $(BOARD_VENDOR_RAMDISK_KERNEL_MODULES), $(wildcard $(TARGET_KERNEL_DIR)/*.ko))
 else
-  BOARD_VENDOR_RAMDISK_KERNEL_MODULES := $(TARGET_MODS)
+  BOARD_VENDOR_KERNEL_MODULES += $(DB845C_ONLY_VENDOR)
+  BOARD_VENDOR_RAMDISK_KERNEL_MODULES := $(filter-out $(DB845C_ONLY_VENDOR),$(TARGET_MODS))
 endif
 
 PRODUCT_SHIPPING_API_LEVEL := 33
diff --git a/full.mk b/full.mk
index 00e3eb8..f445ddf 100644
--- a/full.mk
+++ b/full.mk
@@ -79,8 +79,10 @@ PRODUCT_PACKAGES += \
 #
 # PowerStats HAL
 #
+ifneq ($(POWER_TRACKER_FEATURE), true)
 PRODUCT_PACKAGES += \
     android.hardware.power.stats-service.example
+endif
 
 
 # Audio
diff --git a/shared/utils/init.utils.rc b/shared/utils/init.utils.rc
index a17edfd..3a1867b 100644
--- a/shared/utils/init.utils.rc
+++ b/shared/utils/init.utils.rc
@@ -37,6 +37,8 @@ on early-init
     start set_hw
     exec - root -- /vendor/bin/qrtr-cfg 1
     start qrtr-ns
+    exec u:r:vendor_modprobe:s0 -- /vendor/bin/modprobe -a -d \
+        /vendor/lib/modules nxpnfc_spi
 
 on post-fs
     start pd_mapper
@@ -49,6 +51,7 @@ on post-fs-data
     mkdir /data/vendor/readonly/firmware
     mkdir /data/vendor/readonly/firmware/image
     write /sys/class/firmware/timeout 1
+    write /sys/power/wake_lock nosleep
     start tqftpserv
     start rmtfs
 
diff --git a/ueventd.common.rc b/ueventd.common.rc
index 8ae7d94..9617a38 100644
--- a/ueventd.common.rc
+++ b/ueventd.common.rc
@@ -13,3 +13,7 @@ subsystem usbmisc
 # usb webcam?
 /dev/video0 0660 system camera
 /dev/video1 0660 system camera
+
+/dev/ttyACM0     0666 nfc nfc
+/dev/ttyACM1     0666 nfc nfc
+/dev/nxpnfc     0666 nfc nfc
