diff --git a/db845c.mk b/db845c.mk
index 8df0286..1cf4ffc 100644
--- a/db845c.mk
+++ b/db845c.mk
@@ -10,6 +10,9 @@ $(call inherit-product, device/linaro/dragonboard/shared/graphics/drm_hwcomposer
 $(call inherit-product, device/linaro/dragonboard/shared/graphics/mesa/device.mk)
 $(call inherit-product, device/linaro/dragonboard/shared/graphics/minigbm_msm/device.mk)
 $(call inherit-product, device/linaro/dragonboard/db845c/device.mk)
+$(call inherit-product, vendor/nxp/nfc/device-nfc-vendor.mk)
+$(call inherit-product, system/nxp/PN7xxx/device-nfc-system.mk)
+$(call inherit-product, vendor/nxp/emvco/device-emvco.mk)
 
 #Boot with Android boot image header v3
 ifeq ($(TARGET_KERNEL_USE),5.4)
diff --git a/db845c/BoardConfig.mk b/db845c/BoardConfig.mk
index 7382102..fd3d0d9 100644
--- a/db845c/BoardConfig.mk
+++ b/db845c/BoardConfig.mk
@@ -1,4 +1,5 @@
 include device/linaro/dragonboard/BoardConfigCommon.mk
+include vendor/nxp/nfc/BoardConfigNfc.mk
 
 # Primary Arch
 TARGET_ARCH := arm64
diff --git a/db845c/device.mk b/db845c/device.mk
index 42f4ce5..402952e 100644
--- a/db845c/device.mk
+++ b/db845c/device.mk
@@ -66,10 +66,11 @@ $(call inherit-product-if-exists, vendor/linaro/rb5/$(EXPECTED_LINARO_VENDOR_VER
 ifeq ($(TARGET_USES_LIBPM), true)
 $(call inherit-product, vendor/linaro/shared/$(EXPECTED_LINARO_VENDOR_VERSION)/external/libpm/device.mk)
 else
-PRODUCT_PACKAGES += \
-    android.hardware.power-service.example \
-    android.hardware.power.stats-service.example \
-    com.android.hardware.thermal
+    PRODUCT_PACKAGES += android.hardware.power-service.example
+    ifneq ($(POWER_TRACKER_FEATURE), true)
+	PRODUCT_PACKAGES += android.hardware.power.stats-service.example
+    endif
+    PRODUCT_PACKAGES += com.android.hardware.thermal
 endif
 
 TARGET_HARDWARE := db845c
@@ -82,4 +83,8 @@ TARGET_KERNEL_USE ?= 6.12
 PRODUCT_MAX_PAGE_SIZE_SUPPORTED := 16384
 PRODUCT_NO_BIONIC_PAGE_SIZE_MACRO := true
 
+# Exemption for coredomain entry of tinymix and libdrm.so for resolving compilation error
+# Refer to system/sepolicy/tests/treble_labeling_tests.md for more information
+PRODUCT_SELINUX_TREBLE_LABELING_TRACKING_LIST_FILE := device/linaro/dragonboard/sepolicy/tracking_list.yaml
+
 include device/linaro/dragonboard/device-common.mk
diff --git a/device-common.mk b/device-common.mk
index 67e6d59..cb9b6f3 100644
--- a/device-common.mk
+++ b/device-common.mk
@@ -23,6 +23,7 @@ ifeq ($(TARGET_BOOTS_16K), true)
   endif
 endif
 TARGET_MODS := $(wildcard $(TARGET_KERNEL_DIR)/*.ko)
+DB845C_ONLY_VENDOR := $(wildcard $(TARGET_KERNEL_DIR)/nxpnfc_i2c.ko)
 
 BOARD_DO_NOT_STRIP_VENDOR_RAMDISK_MODULES := true
 BOARD_DO_NOT_STRIP_GENERIC_RAMDISK_MODULES := true
@@ -59,19 +61,25 @@ ifeq ($(TARGET_SDCARD_BOOT), true)
   BOARD_VENDOR_KERNEL_MODULES += $(wildcard $(TARGET_KERNEL_DIR)/ufs*qcom.ko)
 ifeq ($(TARGET_USES_GBL), true)
   BOARD_VENDOR_RAMDISK_KERNEL_MODULES := $(filter-out $(BOARD_VENDOR_KERNEL_MODULES),$(TARGET_MODS))
+  BOARD_VENDOR_RAMDISK_KERNEL_MODULES += $(filter-out $(DB845C_ONLY_VENDOR),$(TARGET_MODS))
 else
   BOARD_GENERIC_RAMDISK_KERNEL_MODULES := $(filter-out $(BOARD_VENDOR_KERNEL_MODULES),$(TARGET_MODS))
+  BOARD_GENERIC_RAMDISK_KERNEL_MODULES += $(filter-out $(DB845C_ONLY_VENDOR),$(TARGET_MODS))
 endif
 else ifeq ($(TARGET_USES_LMP), true)
     include device/linaro/dragonboard/shared/utils/dlkm_loader/vendor.modules.list.mk
     BOARD_VENDOR_KERNEL_MODULES := $(patsubst %,$(TARGET_KERNEL_DIR)/%,$(VENDOR_DLKM_KERNEL_MODULES_LIST))
     BOARD_VENDOR_RAMDISK_KERNEL_MODULES := $(patsubst %,$(TARGET_KERNEL_DIR)/%,$(VENDOR_RAMDISK_KERNEL_MODULES_LIST))
+    BOARD_VENDOR_KERNEL_MODULES += $(DB845C_ONLY_VENDOR)
+    BOARD_VENDOR_RAMDISK_KERNEL_MODULES += $(filter-out $(DB845C_ONLY_VENDOR),$(TARGET_MODS))
     BOARD_SYSTEM_KERNEL_MODULES += $(filter-out $(BOARD_VENDOR_KERNEL_MODULES) $(BOARD_VENDOR_RAMDISK_KERNEL_MODULES), $(wildcard $(TARGET_KERNEL_DIR)/*.ko))
 else
   BOARD_VENDOR_RAMDISK_KERNEL_MODULES := $(filter-out $(BOARD_VENDOR_KERNEL_MODULES) $(BOARD_SYSTEM_KERNEL_MODULES),$(TARGET_MODS))
+  BOARD_VENDOR_KERNEL_MODULES += $(DB845C_ONLY_VENDOR)
+  BOARD_VENDOR_RAMDISK_KERNEL_MODULES := $(filter-out $(DB845C_ONLY_VENDOR),$(TARGET_MODS))
 endif
 
-PRODUCT_SHIPPING_API_LEVEL := 36
+PRODUCT_SHIPPING_API_LEVEL := 37
 
 # Check vendor package version
 # If you need to make changes to the vendor partition,
diff --git a/overlay/packages/modules/Nfc/NfcNci/res/values/config.xml b/overlay/packages/modules/Nfc/NfcNci/res/values/config.xml
new file mode 100644
index 0000000..5f41175
--- /dev/null
+++ b/overlay/packages/modules/Nfc/NfcNci/res/values/config.xml
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright (C) 2019 The Android Open Source Project
+
+     Licensed under the Apache License, Version 2.0 (the "License");
+     you may not use this file except in compliance with the License.
+     You may obtain a copy of the License at
+
+          http://www.apache.org/licenses/LICENSE-2.0
+
+     Unless required by applicable law or agreed to in writing, software
+     distributed under the License is distributed on an "AS IS" BASIS,
+     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+     See the License for the specific language governing permissions and
+     limitations under the License.
+-->
+<resources>
+  <integer name="unknown_tag_polling_delay">-1</integer>
+</resources>
diff --git a/sepolicy/audioserver.te b/sepolicy/audioserver.te
new file mode 100644
index 0000000..eb918e2
--- /dev/null
+++ b/sepolicy/audioserver.te
@@ -0,0 +1,2 @@
+allow audioserver appdomain_tmpfs:file { getattr map read write };
+allow audioserver tmpfs:file { map read write };
diff --git a/sepolicy/file_contexts b/sepolicy/file_contexts
index 463fed3..2bcbfc9 100644
--- a/sepolicy/file_contexts
+++ b/sepolicy/file_contexts
@@ -75,3 +75,4 @@
 /vendor/lib(64)?/libglapi\.so								u:object_r:same_process_hal_file:s0
 /vendor/lib(64)?/libqrtr\.so								u:object_r:same_process_hal_file:s0
 /vendor/lib(64)?/libui\.so								u:object_r:same_process_hal_file:s0
+/vendor/lib(64)?/libaconfig_storage_read_api_cc\.so					u:object_r:same_process_hal_file:s0
diff --git a/sepolicy/hal_audio_default.te b/sepolicy/hal_audio_default.te
index cede442..0e34740 100644
--- a/sepolicy/hal_audio_default.te
+++ b/sepolicy/hal_audio_default.te
@@ -1,2 +1,3 @@
 # audit2allow
 allow hal_audio_default vendor_hw_prop:file { getattr map open read };
+allow hal_audio_default tmpfs:file { read write map };
diff --git a/sepolicy/mediaswcodec.te b/sepolicy/mediaswcodec.te
index 57fb75c..afd96d2 100644
--- a/sepolicy/mediaswcodec.te
+++ b/sepolicy/mediaswcodec.te
@@ -1,2 +1,7 @@
 gpu_access(mediaswcodec)
 allow mediaswcodec gpu_device:chr_file { getattr ioctl map open read write };
+
+allow mediaswcodec system_server_tmpfs:file { getattr map read write };
+allow mediaswcodec tmpfs:file { getattr map read write };
+allow mediaswcodec appdomain_tmpfs:file { map read write };
+allow mediaswcodec vendor_minigbm_debug_prop:file { read };
diff --git a/sepolicy/platform_app.te b/sepolicy/platform_app.te
index 775e964..632bb68 100644
--- a/sepolicy/platform_app.te
+++ b/sepolicy/platform_app.te
@@ -1 +1,4 @@
 gpu_access(platform_app)
+
+allow platform_app audioserver_tmpfs:file { read write };
+allow platform_app tmpfs:file { read write};
diff --git a/sepolicy/system_server.te b/sepolicy/system_server.te
index 5e0af39..f887386 100644
--- a/sepolicy/system_server.te
+++ b/sepolicy/system_server.te
@@ -2,3 +2,5 @@ gpu_access(system_server)
 allow system_server wifi_hal_prop:file {open read getattr map};
 allow system_server vendor_file:dir read;
 allow system_server sysfs_wakeup:file { getattr open read };
+
+allow system_server tmpfs:file { read write };
diff --git a/sepolicy/tracking_list.yaml b/sepolicy/tracking_list.yaml
new file mode 100644
index 0000000..1e2f2a2
--- /dev/null
+++ b/sepolicy/tracking_list.yaml
@@ -0,0 +1,8 @@
+coredomain_in_vendor:
+  - domain: tinymix
+    bug: 1315645
+platform_file_in_vendor_file_contexts:
+  - path: /(system|vendor)/lib(64)?/libdrm.so
+    bug: 1315645
+  - path: /system/bin/tinymix
+    bug: 1315645
diff --git a/shared/utils/init.utils.rc b/shared/utils/init.utils.rc
index 0ec68fb..b45a54d 100644
--- a/shared/utils/init.utils.rc
+++ b/shared/utils/init.utils.rc
@@ -37,6 +37,8 @@ on early-init
     start set_hw
     exec - root -- /vendor/bin/qrtr-cfg 1
     start qrtr-ns
+    exec u:r:vendor_modprobe:s0 -- /vendor/bin/modprobe -a -d \
+    /vendor/lib/modules nxpnfc_i2c
 
 on post-fs
     start pd_mapper
@@ -71,6 +73,7 @@ on post-fs-data
     mkdir /data/vendor/readonly/firmware
     mkdir /data/vendor/readonly/firmware/image
     write /sys/class/firmware/timeout 1
+    write /sys/power/wake_lock nosleep
     start tqftpserv
     start rmtfs
 
diff --git a/ueventd.common.rc b/ueventd.common.rc
index 8ae7d94..9617a38 100644
--- a/ueventd.common.rc
+++ b/ueventd.common.rc
@@ -13,3 +13,7 @@ subsystem usbmisc
 # usb webcam?
 /dev/video0 0660 system camera
 /dev/video1 0660 system camera
+
+/dev/ttyACM0     0666 nfc nfc
+/dev/ttyACM1     0666 nfc nfc
+/dev/nxpnfc     0666 nfc nfc
-- 
2.50.1

